### PR TITLE
Refactor test_recwarn into functional tests instead of testing private implementation

### PR DESCRIPTION
**Pull request by nicoddemus**
Replace tests which test implementation details by functional tests, so we can refactor recwarn module
to simplify it by using the warnings.catch_warnings builtin, available since Python 2.6.

I will open a new PR based upon this one with the refactor itself; I'm opening in two PRs to prove that the behavior has not changed, since I had to convert the tests to functional tests first.

### CODE REVIEW

File: testing/test_recwarn.py [hunk 1]
```diff
@@ -91,18 +94,6 @@ def test_deprecated_call_ret(self):
         ret = pytest.deprecated_call(self.dep, 0)
         assert ret == 42
 
-    def test_deprecated_call_preserves(self):
```
Comment (nicoddemus) [inline testing/test_recwarn.py]: The test is about testing the internal implementation, so I removed it.

File: testing/test_recwarn.py [hunk 2]
```diff
@@ -225,9 +216,6 @@ def test_record(self):
         assert len(record) == 1
         assert str(record[0].message) == "user"
 
-        print(repr(record[0]))
```
Comment (nicoddemus) [inline testing/test_recwarn.py]: This was tested the recent change of the `WarningMessage` object to a `namedtuple`, but this is not part of the API and has never been published, so it is safe to go.

**Comment by nicoddemus** [comment]
Just realized that I can make `WarningsRecorder` subclass `warnings.catch_warnings` (it's an object) so it is possible to keep the exact same interface as before and still do the refactor I want, so changing the tests to functional tests is no longer a requirement for me.

Thoughts?

**Comment by RonnyPfannschmidt** [comment]
i strongly dislike how we turn all kinds of local component tests into system tests, its pretty much worst practice

**Comment by nicoddemus** [comment]
I agree, I was just changing them into functional because my initial take had removed the internal classes; but then afterwards I was able to keep the exact same interface as before, so we can keep the tests as they are right now.

**Pull request closed without merging**